### PR TITLE
Improve the compiler plugin validation to fix the detected issues

### PR DESCRIPTION
**Pull request by kalaiyarasiganeshalingam**
## Purpose
$Subject

Fixes:
https://github.com/ballerina-platform/ballerina-standard-library/issues/3492
https://github.com/ballerina-platform/ballerina-standard-library/issues/3680

## Checklist
- [ ] Linked to an issue
- [ ] Updated the specification
- [ ] Updated the changelog
- [ ] Added tests

### CODE REVIEW

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 1]
```diff
@@ -288,9 +328,14 @@ private void checkAutoIncrementFieldMarkAsKey(SyntaxNodeAnalysisContext ctx, Loc
...
     private void validateEntityAnnotation(SyntaxNodeAnalysisContext ctx, TypeDefinitionNode typeDefinitionNode,
-                                          RecordTypeSymbol recordTypeSymbol) {
+                                          RecordTypeSymbol recordTypeSymbol, DocumentId documentId) {
         Optional<MetadataNode> metadata = typeDefinitionNode.metadata();
         Set<String> tableFields = recordTypeSymbol.fieldDescriptors().keySet();
+        Optional<Path> optional = ctx.currentPackage().project().documentPath(documentId);
+        String path = "";
+        if (optional.isPresent()) {
+            path = optional.get().toString();
+        }
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:338]: Seems to be duplictaed

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistCodeAnalyzer.java [hunk 2]
```diff
@@ -22,13 +22,16 @@
...
 /**
  * Persist Code Analyzer.
  */
 public class PersistCodeAnalyzer extends CodeAnalyzer {
 
     @Override
     public void init(CodeAnalysisContext codeAnalysisContext) {
-        codeAnalysisContext.addSyntaxNodeAnalysisTask(new PersistRecordValidator(), SyntaxKind.TYPE_DEFINITION);
+        codeAnalysisContext.addSyntaxNodeAnalysisTask(new PersistRecordValidator(),
+                List.of(SyntaxKind.TYPE_DEFINITION, SyntaxKind.ENUM_DECLARATION));
```
[the remark this answers is not in the document]
Comment (kalaiyarasiganeshalingam) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistCodeAnalyzer.java:35]: I have checked this model, and it is verified.
```
any a = 5;
@persist:Entity {
    key: ["id"]
}
int a1 = <int>a;
```
Error:
`annotation 'ballerina/persist:0.1.0:Entity' is not allowed on var`

This validation is already included.
https://github.com/ballerina-platform/module-ballerina-persist/pull/74/files#diff-78263bcb99960a46b098ff623741b399de4428b78c35974605e16f24695780f9R22

https://github.com/ballerina-platform/module-ballerina-persist/pull/74/files#diff-e91df391c7c470a52a5109f730b896a6fe3cced450891da072b7f78e0189f88fR84

[the remark this answers is not in the document]
Comment (niveathika) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistCodeAnalyzer.java:35]: You are attaching it to a type definition,
```
any a = 5;
int intType = <@persist {
    autoIncrement: -1
} int>a;
``

[the remark this answers is not in the document]
Comment (kalaiyarasiganeshalingam) [reply compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistCodeAnalyzer.java]: Fixed in [11a1090](https://github.com/ballerina-platform/module-ballerina-persist/pull/74/commits/11a109053e14b5f31db5a980784e96821b8427b7)

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 3]
```diff
@@ -110,15 +116,27 @@ public void perform(SyntaxNodeAnalysisContext ctx) {
...
+            Node node = ctx.node();
+            if (node instanceof EnumDeclarationNode) {
+                EnumDeclarationNode enumDeclarationNode = (EnumDeclarationNode) node;
+                Optional<MetadataNode> metadata = enumDeclarationNode.metadata();
+                metadata.ifPresent(metadataNode -> checkPresenceOfEntityAnnotation(ctx, metadataNode.annotations(),
+                        DiagnosticsCodes.PERSIST_128));
+                return;
+            } else if (node instanceof TypeCastExpressionNode) {
+                NodeList<AnnotationNode> annotations = ((TypeCastExpressionNode) node).typeCastParam().annotations();
+                checkPresenceOfEntityAnnotation(ctx, annotations, DiagnosticsCodes.PERSIST_129);
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:128]: `checkPresenceOfEntityAnnotation` is doing two function,
1. Check for entity annotation
2. If present report diagnostics

IMO it should only do the first function. `isEnitityAnnotationPresent`. Based on the flag return report diagnostics should be called from main function. This makes it easier to grab the logic at this level also reduced passing unnecessary params to function

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 4]
```diff
@@ -141,14 +164,33 @@ public void perform(SyntaxNodeAnalysisContext ctx) {
...
-            reportDiagnosticInfo(ctx, typeDefinitionNode.location(), DiagnosticsCodes.PERSIST_119.getCode(),
-                    DiagnosticsCodes.PERSIST_119.getMessage(), DiagnosticsCodes.PERSIST_119.getSeverity());
+    private void checkPresenceOfEntityAnnotation(SyntaxNodeAnalysisContext ctx, NodeList<AnnotationNode> annotations,
+                                                 DiagnosticsCodes diagnosticsCodes) {
+        for (AnnotationNode annotation : annotations) {
+            if (annotation.annotReference().toSourceCode().trim().equals(Constants.ENTITY)) {
+                reportDiagnosticInfo(ctx, annotation.location(), diagnosticsCodes.getCode(),
+                        diagnosticsCodes.getMessage(),
+                        diagnosticsCodes.getSeverity());
+            }
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:174]: If this is sucessful there should be a break in processing rest of the annotation isnt it? Or is there a possibility of having same annotaion attached twice?

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 5]
```diff
@@ -330,16 +367,34 @@ private void validateEntityAnnotation(SyntaxNodeAnalysisContext ctx, TypeDefinit
...
+        }
+    }
+
+    private String getPath(SyntaxNodeAnalysisContext ctx, DocumentId documentId) {
+        Optional<Path> optional = ctx.currentPackage().project().documentPath(documentId);
+        String path = "";
+        if (optional.isPresent()) {
+            path = optional.get().toString();
+        }
+        return path;
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:382]:
```suggestion
        if (optional.isPresent()) {
            return optional.get().toString();
        }
        return "";
```

File: compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java [hunk 6]
```diff
@@ -285,9 +327,10 @@ private void checkAutoIncrementFieldMarkAsKey(SyntaxNodeAnalysisContext ctx, Loc
     }
 
     private void validateEntityAnnotation(SyntaxNodeAnalysisContext ctx, TypeDefinitionNode typeDefinitionNode,
-                                          RecordTypeSymbol recordTypeSymbol) {
+                                          RecordTypeSymbol recordTypeSymbol, DocumentId documentId) {
```
Comment (niveathika) [inline compiler-plugin/src/main/java/io/ballerina/stdlib/persist/compiler/PersistRecordValidator.java:330]: documentId is passed only to get the path. Cant we process and pass it down? Again whats the use of passing ctx to validate method? Unless there is a need to process it except for diagnostic reporting, it should not be passed down

**Review by niveathika — APPROVED** [review]

**Pull request merged**